pkg/util/bulk/aggregator.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package bulk
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+)
+
+// AggregatorEvent describes an event that can be aggregated and stored by the
+// Aggregator. An AggregatorEvent also implements the tracing.LazyTag interface
+// to render its information on the associated tracing span.
+type AggregatorEvent interface {
+	tracing.LazyTag
+
+	// Clone creates a deep copy of the AggregatorEvent.
+	Clone() AggregatorEvent
+	// Combine combines two AggregatorEvents together.
+	Combine(other AggregatorEvent)
+	// Tag returns a string used to identify the AggregatorEvent.
+	Tag() string
+}
+
+// An Aggregator can be used to aggregate and render AggregatorEvents that are
+// emitted as part of its tracing spans' recording.
+type Aggregator struct {
+	mu struct {
+		syncutil.Mutex
+		// aggregatedEvents is a mapping from the tag identifying the
+		// AggregatorEvent to the running aggregate of the AggregatorEvent.
+		aggregatedEvents map[string]AggregatorEvent
+		// sp is the tracing span managed by the Aggregator.
+		sp *tracing.Span
+	}
+}
+
+// Notify implements the tracing.EventListener interface.
+func (b *Aggregator) Notify(event tracing.Structured) {
+	bulkEvent, ok := event.(AggregatorEvent)
+	if !ok {
+		return
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// If this is the first AggregatorEvent with this tag, set it as a LazyTag on
+	// the associated tracing span. This way the AggregatorEvent will be
+	// dynamically Render()ed everytime we pull the tracing for the associated
+	// span.
+	eventTag := bulkEvent.Tag()
+	if _, ok := b.mu.aggregatedEvents[bulkEvent.Tag()]; !ok {
+		b.mu.aggregatedEvents[eventTag] = bulkEvent.Clone()
```

Comment:
Wonder if we should change the interface for `AggregatorEvent` to be similar to what I've seen elsewhere e.g. in Java's reduce pattern: https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#reduce-T-java.util.function.BinaryOperator-

Instead of `Clone()` we use `Identity()` with `Combine()` here, and maybe having an identity buys us more flexibility in the future.

Reply:
nice idea, check it out!